onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -408,38 +270,62 @@ Status MatMulNBits::ComputeInternal(onnxruntime::webgpu::ComputeContext& context
+    const uint32_t num_N_tile = CEIL_DIV(N, tile_n);
+    const uint32_t num_M_tile = CEIL_DIV(M, tile_m);
 
     MatMulNBitsWideTileProgram program{has_zero_points, tile_m, tile_n, nbits};
     program.SetWorkgroupSize(workgroup_size);
-    program.SetDispatchGroupSize((N + tile_n - 1) / tile_n,
-                                 (M + tile_m - 1) / tile_m,
-                                 batch_count);
-    program.CacheHint("Tile" + std::to_string(tile_m) + "x" + std::to_string(tile_n) + "_Block32");
+    program.SetDispatchGroupSize(num_N_tile, num_M_tile, batch_count);
 
     TensorShape reshaped_a_shape{batch_count, M, K / components_a};
     TensorShape reshaped_b_shape{N, n_blocks_per_col, blob_size_in_words / components_b};
     TensorShape reshaped_y_shape{batch_count, M, N / components};
 
-    program
-        .AddInputs({{a, ProgramTensorMetadataDependency::TypeAndRank, reshaped_a_shape, onnxruntime::narrow<int>(components_a)},
-                    {b, ProgramTensorMetadataDependency::TypeAndRank, reshaped_b_shape, onnxruntime::narrow<int>(components_b * 4)},
-                    {scales, ProgramTensorMetadataDependency::None}})
-        .AddOutput({y, ProgramTensorMetadataDependency::TypeAndRank, reshaped_y_shape, onnxruntime::narrow<int>(components)})
-        .AddUniformVariables({{block_size}, {zero_blocks_per_col}, {num_N_tile}, {num_M_tile}})
-        .CacheHint(nbits, has_zero_points);
+    constexpr uint32_t kU32Components = 4;
+    const uint32_t components_b_with_u32 = components_b * kU32Components;
+    const uint32_t K_of_b = n_blocks_per_col * blob_size / components_b_with_u32;
+
+    program.AddInput({a,
+                      ProgramTensorMetadataDependency::TypeAndRank,
+                      reshaped_a_shape,
+                      onnxruntime::narrow<int>(components_a)});
+    program.AddInput({b,
+                      ProgramTensorMetadataDependency::TypeAndRank,
+                      reshaped_b_shape,
+                      onnxruntime::narrow<int>(components_b_with_u32)});
+    program.AddInput({scales, ProgramTensorMetadataDependency::None});
     if (has_zero_points) {
-      program.AddInput({zero_points, ProgramTensorMetadataDependency::None, {(zero_points->Shape().Size() + 3) / 4}, 4});
+      program.AddInput({zero_points,
+                        ProgramTensorMetadataDependency::TypeAndRank,
+                        {CEIL_DIV(zero_points->Shape().Size(), 4)},
+                        4});
     }
+    program.AddOutput({y,
+                       ProgramTensorMetadataDependency::TypeAndRank,
+                       reshaped_y_shape,
+                       onnxruntime::narrow<int>(components)});
+    program.AddUniformVariables({{batch_count},
+                                 {M},
+                                 {N},
+                                 {K},
+                                 {CEIL_DIV(K, 4)},
+                                 {CEIL_DIV(K, 32)},
```

Comment:
Maybe directly use `n_blocks_per_col` instead of `{CEIL_DIV(K, 32)}`? You can also reuse this name if you want to support other block size.